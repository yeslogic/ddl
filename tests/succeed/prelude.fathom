// Combinators

let id : fun (A : _) -> A -> A
  = fun _ => fun a => a;

let always : fun (A : _) -> fun (B : _) -> A -> B -> A
  = fun _ => fun _ => fun a => fun _ => a;

let compose : fun (A : _) -> fun (B : _) -> fun (C : _) -> (A -> B) -> (B -> C) -> (A -> C)
  = fun _ => fun _ => fun _ => fun ab => fun bc =>
      fun a => bc (ab a);


// Natural numbers

let Nat : Type
  = fun (Nat : _) ->
      fun (succ : Nat -> Nat) ->
      fun (zero : Nat) ->
        Nat;

let zero : Nat
    = fun Nat => fun succ => fun zero => zero;

let succ : Nat -> Nat
    = fun prev => fun Nat => fun succ => fun zero =>
        succ (prev _ succ zero);


let add : Nat -> Nat -> Nat
  = fun n0 => fun n1 => fun Nat => fun succ => fun zero =>
      n0 Nat succ (n1 Nat succ zero);

let mul : Nat -> Nat -> Nat
  = fun n0 => fun n1 =>
      fun Nat => fun succ => fun zero =>
        n0 Nat (n1 Nat succ) zero;


// Lists

let List : Type -> Type
  = fun Elem =>
      fun (List : _) ->
      fun (nil : List) ->
      fun (cons : Elem -> List -> List) ->
        List;

let nil : fun (Elem : _) -> List Elem
  = fun Elem =>
      fun List => fun nil => fun cons => nil;

let cons : fun (Elem : _) -> Elem -> List Elem -> List Elem
  = fun Elem => fun head => fun tail =>
      fun List => fun nil => fun cons =>
        cons head (tail _ nil cons);


// Vectors

let Vec : Type -> Nat -> Type
  = fun Elem => fun len =>
      fun (Vec : Nat -> Type) ->
      fun (nil : Vec zero) ->
      fun (cons : fun (len : _) -> Elem -> Vec len -> Vec (succ len)) ->
        Vec len;

let vnil : fun (Elem : _) -> Vec Elem zero
  = fun Elem =>
      fun Vec => fun nil => fun cons => nil;

let vcons : fun (Elem : _) -> fun (len : _) -> Elem -> Vec Elem len -> Vec Elem (succ len)
  = fun Elem => fun len => fun head => fun tail =>
      fun Vec => fun nil => fun cons =>
        cons _ head (tail Vec nil cons);


// Void

let Void : Type
  = fun (Void : Type) -> Void;

let absurd : fun (A : Type) -> Void -> A
  = fun A => fun void => void A;


// Unit

let Unit : Type
  = fun (Unit : Type) ->
      fun (unit : Unit) ->
        Unit;

let unit : Unit
  = fun Unit => fun unit => unit;


// Leibniz equality

// This states that two things of the same type are equal if every property of
// one is also a property of the other.
let Eq : fun (A : _) -> A -> A -> Type
  = fun A => fun a0 => fun a1 =>
      fun (P : A -> Type) -> P a0 -> P a1;

let refl : fun (A : _) -> fun (a : A) -> Eq A a a
  = fun A => fun a => fun P => id (P a);


let trans : fun (A : _) -> fun (a0 : A) -> fun (a1 : A) -> fun (a2 : A) ->
    Eq _ a0 a1 -> Eq _ a1 a2 -> Eq _ a0 a2
  = fun _ => fun a0 => fun a1 => fun a2 => fun p0 => fun p1 =>
      fun P => compose _ _ _ (p0 P) (p1 P);

let sym : fun (A : _) -> fun (a0 : A) -> fun (a1 : A) -> Eq _ a0 a1 -> Eq _ a1 a0
  = fun _ => fun a0 => fun a1 => fun p =>
      p (fun a1 => Eq _ a1 a0) (refl _ _);


// Examples

let id_apply_type = (fun a => a) Type;

let list1 : List Bool
  = cons _ (id _ true) (nil _);

let five : Nat = succ (succ (succ (succ (succ zero))));
let ten : Nat = add five five;
let hundred : Nat = mul ten ten;
let thousand : Nat = mul ten hundred;

let eq_test : Eq _ hundred hundred = refl _ _;

let eq_id_apply_type : Eq _ ((fun a => a) Type) Type = refl _ _;
let eq_id_apply_true : Eq _ ((fun a => a) true) true = refl _ _;
let eq_id_apply_false : Eq _ ((fun a => a) false) false = refl _ _;


Type
