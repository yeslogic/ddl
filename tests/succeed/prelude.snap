stdout = '''
let id = fun (a : Type) (a : a) => a;
let always = fun (a : Type) (b : Type) (a : a) (_ : b) => a;
let compose = fun (a : Type) (b : Type) (c : Type) (ab : a -> b) (bc : b ->
c) (a : a) => bc (ab a);
let Nat = fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat;
let zero = fun (Nat : Type) (succ : Nat -> Nat) (zero : Nat) => zero;
let succ = fun (prev : fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (Nat :
Type) (succ : Nat -> Nat) (zero : Nat) => succ (prev Nat succ zero);
let add = fun (n0 : fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (n1 :
fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (Nat : Type) (succ : Nat ->
Nat) (zero : Nat) => n0 Nat succ (n1 Nat succ zero);
let mul = fun (n0 : fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (n1 :
fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (Nat : Type) (succ : Nat ->
Nat) (zero : Nat) => n0 Nat (n1 Nat succ) zero;
let List = fun (Elem : Type) => fun (List : Type) -> List -> (Elem -> List ->
List) -> List;
let nil = fun (Elem : Type) (List : Type) (nil : List) (cons : Elem -> List ->
List) => nil;
let cons = fun (Elem : Type) (head : Elem) (tail : fun (List : Type) -> List ->
(Elem -> List -> List) -> List) (List : Type) (nil : List) (cons : Elem -> List
-> List) => cons head (tail List nil cons);
let Vec = fun (Elem : Type) (len : fun (Nat : Type) -> (Nat -> Nat) -> Nat ->
Nat) => fun (Vec : Nat -> Type) -> Vec zero -> (fun (len : fun (Nat : Type) ->
(Nat -> Nat) -> Nat -> Nat) -> Elem -> Vec len -> Vec (succ len)) -> Vec len;
let vnil = fun (Elem : Type) (Vec : (fun (Nat : Type) -> (Nat -> Nat) -> Nat ->
Nat) -> Type) (nil : Vec (fun (Nat : Type) (succ : Nat -> Nat) (zero :
Nat) => zero)) (cons : fun (len : fun (Nat : Type) -> (Nat -> Nat) -> Nat ->
Nat) -> Elem -> Vec len -> Vec (fun (Nat : Type) (succ : Nat -> Nat) (zero :
Nat) => succ (len Nat succ zero))) => nil;
let vcons = fun (Elem : Type) (len : fun (Nat : Type) -> (Nat -> Nat) -> Nat ->
Nat) (head : Elem) (tail : fun (Vec : (fun (Nat : Type) -> (Nat -> Nat) -> Nat
-> Nat) -> Type) -> Vec (fun (Nat : Type) (succ : Nat -> Nat) (zero :
Nat) => zero) -> (fun (len : fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) ->
Elem -> Vec len -> Vec (fun (Nat : Type) (succ : Nat -> Nat) (zero :
Nat) => succ (len Nat succ zero))) -> Vec len) (Vec : (fun (Nat : Type) -> (Nat
-> Nat) -> Nat -> Nat) -> Type) (nil : Vec (fun (Nat : Type) (succ : Nat ->
Nat) (zero : Nat) => zero)) (cons : fun (len : fun (Nat : Type) -> (Nat -> Nat)
-> Nat -> Nat) -> Elem -> Vec len -> Vec (fun (Nat : Type) (succ : Nat ->
Nat) (zero :
Nat) => succ (len Nat succ zero))) => cons len head (tail Vec nil cons);
let Void = fun (Void : Type) -> Void;
let absurd = fun (A : Type) (void : fun (Void : Type) -> Void) => void A;
let Unit = fun (Unit : Type) -> Unit -> Unit;
let unit = fun (Unit : Type) (unit : Unit) => unit;
let Eq = fun (A : Type) (a0 : A) (a1 : A) => fun (P : A -> Type) -> P a0 ->
P a1;
let refl = fun (A : Type) (a : A) (P : A -> Type) => id (P a);
let trans = fun (a : Type) (a0 : a) (a1 : a) (a2 : a) (p0 : fun (P : a ->
Type) -> P a0 -> P a1) (p1 : fun (P : a -> Type) -> P a1 -> P a2) (P : a ->
Type) => compose (P a0) (P a1) (P a2) (p0 P) (p1 P);
let sym = fun (a : Type) (a0 : a) (a1 : a) (p : fun (P : a -> Type) -> P a0 ->
P a1) => p (fun (a1 : a) => Eq a a1 a0) (refl a a0);
let id_apply_type = (fun (a : Type) => a) Type;
let list1 = cons Bool (id Bool true) (nil Bool);
let five = succ (succ (succ (succ (succ zero))));
let ten = add five five;
let hundred = mul ten ten;
let thousand = mul ten hundred;
let eq_test = refl (fun (Nat : Type) -> (Nat -> Nat) -> Nat -> Nat) (fun (_ :
reported_error) (a : reported_error) (b :
reported_error) => a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a (a b))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))));
let eq_id_apply_type = refl Type Type;
let eq_id_apply_true = refl Bool true;
let eq_id_apply_false = refl Bool false;
Type : Type
'''
stderr = ''
